### PR TITLE
[MM-57942] Fix a panic on password is too long

### PR DESCRIPTION
**Pull request by cpoile**
#### Summary
- We were `panic`ing if bcrypt was given a too-long password, probably because we thought all the code paths leading to that point would check for the length of the password -- but we don't in the case of `PreSave()`. For a normal user we check `IsPasswordValid`, but for other paths (e.g., old slack imports, converting a bot, making a guest, ~480 other references, but many are tests...) we don't. Instead of repeating the validation code at each of the calling sites, I added the code once in the `PreSave` -- I think that's cleaner, and protects us in the future if another path saves a user without validating the password.

#### Ticket Link
- https://mattermost.atlassian.net/browse/MM-57942

#### Release Note

```release-note
Fixed panic when password is too long.
```

### CODE REVIEW

File: server/channels/app/user.go [hunk 1]
```diff
@@ -1445,7 +1445,11 @@ func (a *App) UpdatePassword(rctx request.CTX, user *model.User, newPassword str
 		return model.NewAppError("UpdatePassword", "api.user.update_password.failed.app_error", nil, "", http.StatusInternalServerError)
 	}
 
-	hashedPassword := model.HashPassword(newPassword)
+	hashedPassword, err := model.HashPassword(newPassword)
+	if err != nil {
+		// can't be password length (checked in IsPasswordValid)
+		return model.NewAppError("UpdatePassword", "api.user.update_password.password_hash.app_error", nil, "user_id="+user.Id, http.StatusInternalServerError).Wrap(err)
+	}
```
Comment (cpoile) [inline server/channels/app/user.go:1452]: Also handle other bcrypt errors.

File: server/channels/app/users/password.go [hunk 2]
```diff
@@ -20,16 +20,6 @@ func CheckUserPassword(user *model.User, password string) error {
...
-// HashPassword generates a hash using the bcrypt.GenerateFromPassword
-func HashPassword(password string) string {
-	hash, err := bcrypt.GenerateFromPassword([]byte(password), 10)
-	if err != nil {
-		panic(err)
-	}
-
-	return string(hash)
-}
-
```
Comment (cpoile) [inline server/channels/app/users/password.go:32]: Removing dead code. Broken windows policing.

File: server/public/model/user.go [hunk 3]
```diff
@@ -378,10 +379,6 @@ func (u *User) IsValid() *AppError {
 		return InvalidUserError("auth_data_pwd", u.Id, *u.AuthData)
 	}
 
-	if len(u.Password) > UserPasswordMaxLength {
-		return InvalidUserError("password_limit", u.Id, "")
-	}
-
```
Comment (cpoile) [inline server/public/model/user.go:384]: Useless check, at this point u.Password will always be hashed and < 60 bytes.

File: server/i18n/en.json [hunk 4]
```diff
@@ -4142,6 +4142,10 @@
     "id": "api.user.update_password.oauth.app_error",
     "translation": "Update password failed because the user is logged in through an OAuth service."
   },
+  {
+    "id": "api.user.update_password.password_hash.app_error",
+    "translation": ""
```
Comment (cpoile) [inline server/i18n/en.json:4147]: Is this something I need to write myself?

File: server/channels/app/user.go [hunk 5]
```diff
@@ -1445,7 +1445,11 @@ func (a *App) UpdatePassword(rctx request.CTX, user *model.User, newPassword str
 		return model.NewAppError("UpdatePassword", "api.user.update_password.failed.app_error", nil, "", http.StatusInternalServerError)
 	}
 
-	hashedPassword := model.HashPassword(newPassword)
+	hashedPassword, err := model.HashPassword(newPassword)
+	if err != nil {
+		// can't be password length (checked in IsPasswordValid)
```
Comment (streamer45) [inline server/channels/app/user.go:1450]: Can you point me to where we validate the password for a new user? (i.e. signing up)

**Review by streamer45 — COMMENTED** [review]
Thanks. Just a couple of questions.

File: server/i18n/en.json [hunk 6]
```diff
@@ -9754,6 +9758,14 @@
     "id": "model.user.is_valid.username.app_error",
     "translation": "Username must begin with a letter, and contain between 3 to 22 lowercase characters made up of numbers, letters, and the symbols \".\", \"-\", and \"_\"."
   },
+  {
+    "id": "model.user.pre_save.password_hash.app_error",
+    "translation": "There was an internal error saving the password."
+  },
+  {
+    "id": "model.user.pre_save.password_too_long.app_error",
+    "translation": "Your password must contain no more than 72 characters."
```
[the remark this answers is not in the document]
Comment (cpoile) [reply server/i18n/en.json:9767]: It's in constants in a couple places, alluded to in the docs: https://docs.mattermost.com/configure/authentication-configuration-settings.html#minimum-password-length ,  and explicit in the UI: 
image

File: server/channels/app/user.go (hunk 5 above)
Comment (cpoile) [reply server/channels/app/user.go:1450]: https://github.com/mattermost/mattermost/blob/master/server/channels/app/users/users.go#L63

**Review by streamer45 — APPROVED** [review]
Thanks, looks good!

File: server/channels/app/users/password_test.go [hunk 7]
```diff
@@ -13,13 +13,6 @@ import (
 	"github.com/mattermost/mattermost/server/public/model"
 )
 
-func TestComparePassword(t *testing.T) {
-	hash := HashPassword("Test")
```
Comment (enzowritescode) [inline server/channels/app/users/password_test.go:17]: How come you're getting rid of this?

File: server/public/model/user.go [hunk 8]
```diff
@@ -477,14 +475,24 @@ func (u *User) PreSave() {
 	}
 
 	if u.Password != "" {
-		u.Password = HashPassword(u.Password)
+		hashed, err := HashPassword(u.Password)
+		if errors.Is(err, bcrypt.ErrPasswordTooLong) {
+			return NewAppError("User.PreSave", "model.user.pre_save.password_too_long.app_error",
+				nil, "user_id="+u.Id, http.StatusBadRequest).Wrap(err)
```
Comment (enzowritescode) [inline server/public/model/user.go:481]: This feels very weird to have an HTTP error being throw in the model layer. I feel like it should only ever be thrown in the API layer. Or is this just some legacy thing we have to accept when using `AppError`?

File: server/channels/app/users/password_test.go (hunk 7 above)
Comment (cpoile) [reply server/channels/app/users/password_test.go:17]: @enzowritescode That `HashPassword` was never used anywhere, so removing the dead code.

**Review by enzowritescode — APPROVED** [review]

File: server/public/model/user.go [hunk 9]
```diff
@@ -430,7 +428,7 @@ func NormalizeEmail(email string) string {
 // PreSave will set the Id and Username if missing.  It will also fill
 // in the CreateAt, UpdateAt times.  It will also hash the password.  It should
 // be run before saving the user to the db.
-func (u *User) PreSave() {
+func (u *User) PreSave() *AppError {
```
[the remark this answers is not in the document]
Comment (agnivade) [reply server/public/model/user.go:431]: The `model` package is now under 0.x versioning scheme. So there's no fear of breaking change. I'd love to move away from AppError, but we need a plan for that.

**Pull request merged**